### ISSUE TITLE
torch.nn.functional.conv2d produces different results on CPU and GPU

### ISSUE THREAD

**Issue by yypurpose** [pytorch/pytorch#70589]
### 🐛 Describe the bug

I found the online evaluation results (evaluate when training) of my model did not match perfectly with the offline evaluation results (evaluate the saved checkpoint). After rounds of checking, a strange phenomenon appeared, that is, the computation results of the same conv2d function on CPU and GPU do not match. Here is an example.
Codes:
 ```
import numpy as np
import torch
import torch.nn.functional as F

np.random.seed(0)
torch.manual_seed(0)
torch.cuda.manual_seed_all(0)

inputs = torch.randn(size=(64, 3, 32, 32))
weight = torch.randn(size=(16, 3, 3, 3))

results_gpu = F.conv2d(input=inputs.cuda(), weight=weight.cuda(), stride=1, padding=1).cpu()
results_cpu = F.conv2d(input=inputs.cpu(), weight=weight.cpu(), stride=1, padding=1)

print((results_cpu == results_gpu).sum())
print((results_cpu != results_gpu).sum())
print((torch.abs(results_cpu - results_gpu) > 1e-5).sum())
print((torch.abs(results_cpu - results_gpu) > 1e-6).sum())
print((torch.abs(results_cpu - results_gpu) > 1e-7).sum())
print((torch.abs(results_cpu - results_gpu) > 1e-8).sum())
```

Output results:
```
tensor(69622)
tensor(978954)
tensor(33261)
tensor(604133)
tensor(972968)
tensor(978863)
```

It seems like there is a large difference between the results computed on CPU and GPU. Is it acceptable?

### Versions

Collecting environment information...
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.1 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.1) 5.4.0 20160609
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.9

Python version: 3.6.10 |Anaconda, Inc.| (default, May  8 2020, 02:54:21)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.4.0-31-generic-x86_64-with-debian-stretch-sid
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti

Nvidia driver version: 460.67
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.6.0
[pip3] torch-geometric==1.6.2
[pip3] torchcam==0.1.2
[pip3] torchcontrib==0.0.2
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.7.0
[conda] blas                      1.0                         mkl    defaults
[conda] cudatoolkit               10.2.89              hfd86e86_1    defaults
[conda] mkl                       2020.2                      256    defaults
[conda] mkl-service               2.3.0            py36he904b0f_0    defaults
[conda] mkl_fft                   1.1.0            py36h23d657b_0    defaults
[conda] mkl_random                1.1.1            py36h0573a6f_0    defaults
[conda] numpy                     1.19.1           py36hbc911f0_0    defaults
[conda] numpy-base                1.19.1           py36hfa32c7d_0    defaults
[conda] pytorch                   1.6.0           py3.6_cuda10.2.89_cudnn7.6.5_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
[conda] torch-geometric           1.6.2                    pypi_0    pypi
[conda] torchcam                  0.1.2                    pypi_0    pypi
[conda] torchcontrib              0.0.2                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.7.0                py36_cu102    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch

**Comment by qiliux** [comment]
try cloning the input first and see whether there is a difference.
```
results_gpu = F.conv2d(input=inputs.clone().cuda(), weight=weight.cuda(), stride=1, padding=1).cpu()
results_cpu = F.conv2d(input=inputs.clone().cpu(), weight=weight.cpu(), stride=1, padding=1)

```

**Comment by yypurpose** [comment]
> try cloning the input first and see whether there is a difference.
> 
> ```
> results_gpu = F.conv2d(input=inputs.clone().cuda(), weight=weight.cuda(), stride=1, padding=1).cpu()
> results_cpu = F.conv2d(input=inputs.clone().cpu(), weight=weight.cpu(), stride=1, padding=1)
> ```

Hi @qiliux, thanks for your reply. I tried to clone (or even detach) the input, and the output difference is consistent.

**Comment by Lezcano** [comment]
The algorithms to compute the convolution in CPU and CUDA are different, so some differences are expected. Running 
```python
import numpy as np
import torch
import torch.nn.functional as F

np.random.seed(0)
torch.manual_seed(0)
torch.cuda.manual_seed_all(0)

inputs = torch.randn(size=(64, 3, 32, 32))
weight = torch.randn(size=(16, 3, 3, 3))

results_gpu = F.conv2d(input=inputs.cuda(), weight=weight.cuda(), stride=1, padding=1).cpu()
results_cpu = F.conv2d(input=inputs.cpu(), weight=weight.cpu(), stride=1, padding=1)
print((results_gpu - results_cpu).abs().max())
```
in my box outputs `tensor(7.6294e-06)`, which is rather low considering that this is single floating point precision. When using `float64` (**much** slower on GPU but more precise), I get `tensor(3.5527e-15, dtype=torch.float64)` and `print(torch.allclose(results_gpu, results_cpu))` outputs `True`.
